File: typescript/nucleotide-count/nucleotide-count.ts
export interface Histogram {
  A: number;
  C: number;
  G: number;
  T: number;
}

export default class NucleotideCounts {
  static nucleotideCounts(strand: string): Histogram {
    if (!/^[ACGT]*$/.test(strand)) {
      throw new Error('Invalid nucleotide in strand');
    }

    const adenineCount = (strand.match(/[A]/g) || []).length;
    const cytosineCount = (strand.match(/[C]/g) || []).length;
    const guanineCount = (strand.match(/[G]/g) || []).length;
    const thymineCount = (strand.match(/[T]/g) || []).length;

    return {A: adenineCount, C: cytosineCount, G: guanineCount, T: thymineCount};
  }
}

File: go/nucleotide-count/nucleotide_count.go
package dna

import "errors"

// Histogram is a mapping from nucleotide to its count in given DNA.
type Histogram map[rune]int

// DNA is a list of nucleotides.
type DNA string

// Counts generates a histogram of valid nucleotides in the given DNA.
func (d DNA) Counts() (Histogram, error) {
	var h Histogram = Histogram{'A': 0, 'C': 0, 'G': 0, 'T': 0}
	for _, char := range d {
		switch char {
		case 'A':
			h['A']++
			break
		case 'C':
			h['C']++
			break
		case 'G':
			h['G']++
			break
		case 'T':
			h['T']++
			break
		default:
			return Histogram{}, errors.New("Invalid DNA strand")
		}
	}
	return h, nil
}
